### PR TITLE
Feature: Node integration

### PR DESCRIPTION
**Pull request by ghost**
Fixes punctual things in code that allow bundler-node validator integrations.

### CODE REVIEW

File: src/bin/validator.rs [hunk 1]
```diff
@@ -106,13 +105,10 @@ impl From<&CliOpts> for AppContext {
...
-                public: RsaPublic {
-                    e: PublicExponent,
-                    n: n.as_bytes().into(),
-                },
-                private: None,
-            })
+            let jwt_str = format!("{{\"kty\":\"RSA\",\"e\":\"AQAB\",\"n\":\"{}\"}}", n);
+            let jwk: JsonWebKey = jwt_str.parse().unwrap();
+            let p = serde_json::to_string(&jwk).unwrap();
+            p.parse().unwrap()
```
Comment (mikonieminen) [inline src/bin/validator.rs]: Why to switch form building the object into parsing it from a string?

At least, rename to `jwt_str` to `jwk_str`.

File: src/cron/bundle.rs [hunk 2]
```diff
@@ -278,12 +277,11 @@ fn verify_tx_receipt(tx_receipt: &TxReceipt) -> std::io::Result<bool> {
...
-                kty: "RSA",
-                e: "AQAB",
-                n: std::env::var("BUNDLER_PUBLIC").unwrap(),
-            };
-
+            let jwt_str = format!(
+                "{{\"kty\":\"RSA\",\"e\":\"AQAB\",\"n\":\"{}\"}}",
+                std::env::var("BUNDLER_PUBLIC").unwrap()
+            );
+            let jwk: JsonWebKey = jwt_str.parse().unwrap();
```
Comment (mikonieminen) [inline src/cron/bundle.rs]: Same as earlier, why to switch from building the object into creating string representing it and then parsing the it from the string? I don't really see any added value here.

And again, not `jwt_str`, but it should be `jwk_str`.

File: src/server/routes/sign.rs [hunk 3]
```diff
@@ -81,12 +81,9 @@ impl SignRequest {
             error!("Failed to build data for signing: {:?}", err);
         })?;
 
-        let decoded_signature =
-            BASE64URL_NOPAD
-                .decode(self.signature.as_bytes())
-                .map_err(|err| {
-                    error!("Failed to decode signature: {:?}", err);
-                })?;
+        let decoded_signature = BASE64.decode(self.signature.as_bytes()).map_err(|err| {
```
Comment (mikonieminen) [inline src/server/routes/sign.rs]: Is this right? Earlier @joshbenaron said this should be `BASE64URL_NOPAD`.

**Review by mikonieminen — CHANGES_REQUESTED** [review]

File: src/server/routes/sign.rs (hunk 3 above)
Comment (ghost) [reply src/server/routes/sign.rs]: Testing with bundlr-node sending signatures, it does not decode/encode correctly with `BASE64URL_NOPAD`

File: src/cron/bundle.rs (hunk 2 above)
Comment (ghost) [reply src/cron/bundle.rs]: Its being parsed incorrectly. As it was done before, parsing as bytes, it was not generating the jwk correctly.

File: src/server/routes/sign.rs (hunk 3 above)
Comment (ghost) [reply src/server/routes/sign.rs]: After some trial and error, I was able to get it with just `BASE64`

File: src/cron/bundle.rs (hunk 2 above)
Comment (mikonieminen) [reply src/cron/bundle.rs]: Actually, the mistake is that the value for `n` should be decoded first. This is my mistake in the existing code. Creating JSON string and then parsing it, is an unnecessary step. Instead the code should look like:
```rust
let jwk = JWK {
    kty: "RSA",
    e: "AQAB",
    n: BASE64URL_NOPAD.decode(std::env::var("BUNDLER_PUBLIC").unwrap()),
};
```
Though this code should anyways be changed as we should not read the value in a `lazy_static!`, instead we should be using `KeyManager` where needed.

File: src/server/routes/sign.rs (hunk 3 above)
Comment (mikonieminen) [reply src/server/routes/sign.rs]: In https://docs.arweave.org/developers/server/http-api#transaction-format, it's defined to be `BASE64URL`. Maybe it's just not `BASE64URL_NOPAD`

File: src/cron/bundle.rs (hunk 2 above)
Comment (ghost) [reply src/cron/bundle.rs]: Alright, these changes should be made in this PR?

Comment (joshbenaron) [reply src/cron/bundle.rs]: I'm inclined to leave as is for this PR as it works. Create a new PR making this above change. @bluekirby1111 - speak to @mikonieminen if it's unclear as to what you need to do

Comment (joshbenaron) [reply src/cron/bundle.rs]: @bluekirby1111 please read all comments carefully. Miko raised this. Change `jwt_str` to `jwk_str`. JWT !== JWK

File: src/server/routes/sign.rs (hunk 3 above)
Comment (joshbenaron) [reply src/server/routes/sign.rs]: @gustavocalonico can you change BASE64 to BASE64URL_NOPAD. Can you also make PR to change [this](https://github.com/Bundlr-Network/bundler-node/blob/master/src/server/routes/data/postData.ts#L395) to use `base64url` please

Comment (ghost) [reply src/server/routes/sign.rs]: Alright, 2 PRS then

**Comment by ghost** [comment]
Relates to:

https://github.com/Bundlr-Network/bundler-node/pull/81

**Comment by mikonieminen** [comment]
If we can first review and merge #32 and #30 (in that order) and I will rebase and do a final review once those two are in master.

File: src/bin/validator.rs [hunk 4]
```diff
@@ -38,7 +38,7 @@ struct CliOpts {
     database_url: String,
 
     /// Listen address for the server
-    #[clap(short, long, env, default_value = "127.0.0.1:10000")]
+    #[clap(short, long, env, default_value = "127.0.0.1:1234")]
```
Comment (joshbenaron) [inline src/bin/validator.rs:41]: How comes this was changed?

File: src/server/routes/sign.rs (hunk 3 above)
Comment (joshbenaron) [reply src/server/routes/sign.rs]: Has the second PR been created?

**Comment by mikonieminen** [comment]
@bluekirby1111 I think we should close this PR or does this have any purpose?

File: src/bin/validator.rs (hunk 4 above)
Comment (ghost) [reply src/bin/validator.rs:41]: The bundlr-node default port is already 10000, so I commited this.

**Comment by ghost** [comment]
Yes, should be closed. Your PR before from this one solved all issues.

**Pull request closed without merging**